accepted/future-releases/nnbd/number-operation-typing.md

```diff
@@ -62,47 +62,76 @@ T add<T extends num>(T a, T b) => a + b;
 
 because any attempt to plug in actual values will give sound results.
 
-Users also understand that `clamp` will return either the receiver or one of the arguments. If those all have the same type, the result will have that type.
+Users also understand that `clamp` will return either the receiver or one of the arguments. If those all have the same type, the result will have that type.
 
 ## Solution
 
 ### Extended Rules
 
-We extend the special-casing rules of `+`, `-`, `*` and `%` to also cover calls of the `remainder` method, and to also work with type parameters which extend `num` , `int` or `double`. Finally, if the second operand is a `double` and the first is a `num`, the result is guaranteed to be a `double`.
+We extend the special-casing rules of `+`, `-`, `*` and `%` to also cover calls of the `remainder` method, and to also work with type parameters which extend `num` , `int` or `double`. Finally, if the second operand is a `double` and the first is a `num`, the result is guaranteed to be a `double`.
 
 That is:
 
-> For an expression `e` of one of the forms `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)`, where the static type of `e1` is a non-`Never` type *T* where *T* <: `num` and the static type of `e2` is a non-`Never` type *S* where *S* <: `num`:
+> For an expression `e` of one of the forms `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)`, where the static type of `e1` is a non-`Never` type *T* where *T* <: `num` and the static type of `e2` is a non-`Never` type *S* where *S* <: `num`:
 >
-> * If S <: T then the static type of `e` is *T*. 
-> * Otherwise If *T* <: *S* then the static type of `e` is *S*.
-> * Otherwise if *T* <: `int` and *S* <: `int` then the static type of `e` is `int`.
-> * Otherwise if *T* <: `double` or *S* <: *double* then the static type of *e* is `double`.
+> * If *T* <: `int` and *S* <: `int` then the static type of `e` is *T*.
+> * Otherwise if *T* <: `double` , then the static type of `e` is *T*.
+> * Otherwise if *S* <: `double` then the static type of `e` is `double`.
+> * Otherwise if *T* is not a subtype of `int` or `double`, then the static type of `e` is *T*.
```

Comment:
Ack, so I though of that the first time around.

But LUB is not what I want here, because that doesn't work.

Say x has type *T* = `X&int` and y has type *S* = `int`, then `x += y` should be valid. That only works if we take the type of *T* for the int+int case and the double+anything case.

I guess the rule should be:

* If *S* is a non-`Never` subtype of `int`, then the static type of `e` is *T*.  *(Anything + int is the first thing again, true and sound for all of `int`, `double` and `num`)*.
* Otherwise, if `T` is a non-`Never` subtype of `double`, then the static type of `e` is *T*. *(`Double` + anything is the same double again)*
* Otherwise, if *S* is a non-`Never` subtype of `double`, then the static type of `e` is `double`. *(Anything + `double` is `double`).
* Otherwise the static type of `e` is `num`.

That also fits better with how I'm thinking about it.

Reply:
The new rules make sense to me!

Reply:
It does miss the part where "if *S* and *T* is the same type, there result is *T*." which allows `T add<T extends num>(T a, T b) => a + b;` to work.
Maybe it needs to be phrased as:

> * Otherwise, If *S* <: *T* and *T* <: *S*, the static type of `e` is *T*.
> * Otherwise, the static type of `e` is `num`.

Reply:
Yep!